### PR TITLE
fix: stop a provider-refused parameter from making a whole model unusable

### PR DESCRIPTION
**Pull request by varunursekar**
Fixes the `opencode x kimi-k3` grid cell, which could not run at all.

## The failure

`opencode`s first tool-bearing request came back 400:

```
litellm.UnsupportedParamsError: fireworks_ai does not support parameters:
[tool_choice], for model=accounts/fireworks/models/kimi-k3
```

The agent exited non-zero and the trial died 2m 04s in with no candidate. The
model is not the problem: with the parameter absent it emits tool calls normally.
One parameter the provider declines to accept was blocking an entire contestant.

## Why the documented workarounds do not cover it

The upstream error suggests three fixes. Two configure the upstream proxy, which
we do not own. The third is `allowed_openai_params`, and it turns out to be
endpoint dependent. Measured against the live upstream:

| request | `/chat/completions` | `/responses` |
|---|---|---|
| tools only | 200 | 200 |
| `tool_choice` | 400 | 400 |
| `tool_choice` + `allowed_openai_params` | **200** | **400** |

An openai-prefixed harness talks to `/responses`, which is exactly where the
escape hatch is silently ignored. I shipped the `allowed_openai_params` version
first and it passed a hand-rolled end-to-end check on `/chat/completions` while
the real conformance run still failed on `/responses` — the request log showed the
parameter being added and refused anyway.

## The fix

The gateway reads the offending parameter names out of the 400 and sends once
more without them. A refused parameter fails the request before it reaches the
model, so nothing has been billed and nothing has been streamed downstream; the
retry is safe at that point and nowhere later.

**Retrying rather than stripping is deliberate.** Only a request that has *already*
failed is touched, so providers that accept `tool_choice` keep their exact
semantics. An unconditional strip would have silently changed behaviour for every
harness that relies on it, which is not a trade worth making mid-experiment.
Dropped parameters are recorded in the request log, so a degraded request is
auditable rather than invisible.

## Verification

A full `harness-conformance` run with `opencode x fireworks_ai/kimi-k3`:

- **reward 1.0**, level with `codex` and `claude-code` on the same target
- `error_rate` 0.0, `shipped: true`, 9m 40s (the failing run died at 1m 38s)
- zero `UnsupportedParamsError` in the agent log
- retry fired on **52 of 66** gateway requests, every one 200

Tests cover the retry on `/responses`, no retry when the provider accepts the
parameter, and no retry on a 400 that blames nothing droppable. Full suite: 483
passed, 15 skipped.

## Follow-up, not in scope here

`opencode` runs a separate small model for session-title generation that is
absent from the gateway scope allow-list, so every run logs an avoidable 403
(1 of the 66 requests above). Non-fatal — the r1 opencode cells scored normally
with it — but worth cleaning up separately.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<h3>Greptile Summary</h3>

This PR fixes a provider-refused parameter (`tool_choice`) that was making the entire `kimi-k3` model unusable through the gateway. On a 400 response naming unsupported parameters, the gateway now strips only those parameters and retries once; it also caches the refusal per-model so subsequent requests are pre-stripped without an extra round-trip.

- **Retry logic** (`inference.py`): A new `_unsupported_params` helper parses the error body (preferring `error.message` over raw bytes to avoid JSON-escaping issues) and extracts blamed parameter names. On a 400, the gateway strips matching parameters, re-serializes the body, records `dropped_params` in the request log for auditability, persists the refusal in `app.state.refused_params`, and sends once more. The retry is gated on a request that has already failed, so providers that accept the parameter keep their exact semantics.
- **Tests** (`test_v05_harbor_inference.py`): Five new tests cover the retry on `/responses`, no retry on success, no retry for unrelated 400s, JSON-escaped quoting edge cases in the error parser, the `dropped_params` audit field, and the one-discovery-per-model efficiency property.

<details><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge — the retry is narrowly scoped to requests that have already returned 400, leaves all successful paths untouched, and is fully covered by new tests including the audit-trail regression pin.

The retry logic engages only after a confirmed upstream refusal, so it cannot affect providers that accept the parameter. The refused_params cache is bounded by the gateway's allowed_models set and is never written without a provider-confirmed 400. Tests cover the retry path, the no-retry path, the audit field, and the efficiency property.

**Files Needing Attention:** No files require special attention.
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| vero/src/vero/gateway/inference.py | Adds _unsupported_params helper, per-model refused_params cache, and a single-retry loop that strips provider-blamed parameters; the retry only fires on an already-failed 400, leaving successful providers untouched. |
| vero/tests/test_v05_harbor_inference.py | Five new test cases cover the retry on /responses, no-retry on success, no-retry for unrelated 400s, JSON-escaped quoting in the error parser, the dropped_params audit field, and the one-discovery-per-model efficiency guarantee. |

</details>

<sub>Reviews (3): Last reviewed commit: ["Learn a provider&#39;s refusal once, not on ..."](https://github.com/scaleapi/vero/commit/aa3ba2993a6efe1d7f9c92050c1a07c121f75106) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=48573060)</sub>

### CODE REVIEW

File: vero/src/vero/gateway/inference.py [hunk 1]
```diff
@@ -666,6 +664,42 @@ def _cost_fields(headers: Any) -> dict[str, float]:
...
+    Retrying without the blamed parameters is, and it is the conservative option:
+    it engages only on a request that has *already* failed, so providers that
+    accept the parameter are untouched and keep their exact semantics. That
+    matters more than elegance here -- an unconditional strip would silently
+    change behaviour for every harness that relies on ``tool_choice``.
+    """
+    match = _UNSUPPORTED_PARAMS.search(body.decode("utf-8", "replace"))
+    if not match:
+        return []
+    return re.findall(r"['\"]([^'\"]+)['\"]", match.group(1))
```
[the remark this answers is not in the document]
Comment (varunursekar) [reply vero/src/vero/gateway/inference.py]: Valid, fixed in 7fb15be — now parses the body and matches `error.message`, with a raw-body fallback so an unexpected shape degrades to best effort rather than giving up.

One correction to the mechanism, since the distinction matters for the regression test: it does not fail to match. Against the escaped body the outer pattern captures the quoted element, then the inner character class `[^'"]+` consumes `tool_choice\` — backslash is not excluded by that class — before the closing quote, so it returns `tool_choice\` with a trailing backslash. The name then fails the `name in value` check, so your conclusion is right (the retry silently never fires) but via a trailing-backslash capture rather than a non-match. I verified this before changing anything, and `test_unsupported_params_survives_json_escaped_quoting` pins the real behaviour for both quotings, multiple parameters, an unrelated error, and a non-JSON body.

File: vero/tests/test_v05_harbor_inference.py [hunk 2]
```diff
@@ -862,3 +864,108 @@ def upstream(request: httpx.Request):
...
+    )
+    with TestClient(app) as client:
+        failed = client.post(
+            "/scopes/producer/optimizer/v1/responses",
+            headers={"Authorization": "Bearer scoped-token"},
+            json={"model": "gpt-test", "tool_choice": "auto", "input": []},
+        )
+
+    assert failed.status_code == 400
+    assert len(seen) == 1
```
[the remark this answers is not in the document]
Comment (varunursekar) [reply vero/tests/test_v05_harbor_inference.py:973]: Valid, fixed in 7fb15be. Added `test_gateway_records_dropped_params_in_the_request_log`, which wires a real `InferenceRequestLogConfig` and asserts the retried record carries `dropped_params: ["tool_choice"]`.

It also asserts the negative: a request nothing was dropped from must have no `dropped_params` key rather than an empty list, so the field cannot silently become noise on every untouched request. Worth pinning because the retry changes what was asked of the provider without the caller knowing — if the field regresses, every future degraded request becomes invisible and the only remaining signal is a behaviour difference nobody is watching for.
